### PR TITLE
Feature/dev 533 bank account api

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/payment/models/bank-account/bank-account.controller.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    @GetJwt() jwt: JwtPayload,
+    @Body() createBankAccountDto: CreateBankAccountDto,
+  ): Promise<BankAccountDto> {
+    return this.bankAccountService.createBankAccount(jwt.id, createBankAccountDto).then((b) => this.toDto(b));
+  }
+
+  @Put()
+  @ApiBearerAuth()
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.USER))
+  async updateBuyRoute(
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.controller.ts:40]: Rename the methods

File: src/payment/models/bank-account/bank-account.service.ts [hunk 2]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
 
     return (
-      bankAccounts.find((b) => b.user.id === userId) ?? (await this.createBankAccount(iban, userId, bankAccounts[0]))
+      bankAccounts.find((b) => b.user.id === userId) ??
+      (await this.createBankAccountInternal(iban, userId, bankAccounts[0]))
     );
   }
 
-  private async createBankAccount(iban: string, userId: number, copyFrom?: BankAccount): Promise<BankAccount> {
+  //
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:29]: Empty comment? 😉

File: src/payment/models/bank-account/bank-account.service.ts [hunk 3]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+  async getUserBankAccounts(userId: number): Promise<BankAccount[]> {
+    return this.bankAccountRepo.find({ user: { id: userId } });
+  }
+
+  async createBankAccount(userId: number, dto: CreateBankAccountDto): Promise<BankAccount> {
+    const bankAccounts = await this.bankAccountRepo.findOne({
+      where: { iban: dto.iban, user: { id: userId } },
+      relations: ['user'],
+    });
+    if (bankAccounts) throw new ForbiddenException('bankAccount already exists');
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:40]:
```ts
if (bankAccounts) throw new ConflictException('BankAccount already exists');
```

File: src/payment/models/bank-account/bank-account.service.ts [hunk 4]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+
+    const bankAccount = await this.getBankAccount(dto.iban, userId);
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:42]: Rename `getBankAccount ` to `getOrCreateBankAccount`

File: src/payment/models/bank-account/bank-account.service.ts [hunk 5]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+
+    bankAccount.user = { id: userId } as User;
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:44]: This line can be removed

File: src/payment/models/bank-account/bank-account.service.ts [hunk 6]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+    bankAccount.preferredCurrency = { id: dto.preferredCurrency.id } as Fiat;
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:45]:
```ts
  // check currency
  if (dto.preferredCurrency) {
    bankAccount.preferredCurrency = await this.fiatService.getFiat(dto.preferredCurrency.id);
    if (!bankAccount.preferredCurrency) throw new BadRequestException('Currency not found');
  }
```

File: src/payment/models/bank-account/bank-account.service.ts [hunk 7]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
 
-  private async createBankAccount(iban: string, userId: number, copyFrom?: BankAccount): Promise<BankAccount> {
+  //
+
+  async getUserBankAccounts(userId: number): Promise<BankAccount[]> {
+    return this.bankAccountRepo.find({ user: { id: userId } });
+  }
+
+  async createBankAccount(userId: number, dto: CreateBankAccountDto): Promise<BankAccount> {
+    const bankAccounts = await this.bankAccountRepo.findOne({
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:36]: Rename `bankAccounts` to `existing`

File: src/payment/models/bank-account/bank-account.service.ts [hunk 8]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+      where: { iban: dto.iban, user: { id: userId } },
+      relations: ['user'],
+    });
+    if (bankAccounts) throw new ForbiddenException('bankAccount already exists');
+
+    const bankAccount = await this.getBankAccount(dto.iban, userId);
+
+    bankAccount.user = { id: userId } as User;
+    bankAccount.preferredCurrency = { id: dto.preferredCurrency.id } as Fiat;
+    if (dto.label) bankAccount.label = dto.label;
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:46]: `if (dto.label)` can be removed IMO

File: src/payment/models/bank-account/bank-account.service.ts [hunk 9]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+
+    return this.bankAccountRepo.save(bankAccount);
+  }
+
+  async updateBankAccount(userId: number, dto: UpdateBankAccountDto): Promise<BankAccount> {
+    let entity = await this.bankAccountRepo.findOne({
+      where: { iban: dto.iban, user: { id: userId } },
+      relations: ['user'],
+    });
+    if (!entity) throw new NotFoundException('bankAccount not found');
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:56]: **B**ankAccount not found

File: src/payment/models/bank-account/dto/create-bank-account.dto.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  label: string;
+
+  @ApiProperty()
+  @IsNotEmptyObject()
+  @ValidateNested()
+  @Type(() => EntityDto)
+  preferredCurrency: Fiat;
```
Comment (davidleomay) [inline src/payment/models/bank-account/dto/create-bank-account.dto.ts:20]: Optional?

File: src/payment/models/bank-account/dto/create-bank-account.dto.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+import { Type } from 'class-transformer';
+import { IsNotEmpty, IsNotEmptyObject, IsOptional, ValidateNested } from 'class-validator';
+import { EntityDto } from 'src/shared/dto/entity.dto';
+import { Fiat } from 'src/shared/models/fiat/fiat.entity';
+
+export class CreateBankAccountDto {
+  @ApiProperty()
+  @IsNotEmpty()
+  iban: string;
```
Comment (davidleomay) [inline src/payment/models/bank-account/dto/create-bank-account.dto.ts:10]: Missing @IsString()

File: src/payment/models/bank-account/dto/create-bank-account.dto.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+  @ApiPropertyOptional()
+  @IsOptional()
+  label: string;
```
Comment (davidleomay) [inline src/payment/models/bank-account/dto/create-bank-account.dto.ts:14]: Missing @IsString()

File: src/payment/models/bank-account/dto/update-bank-account.dto.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+import { IsNotEmpty, IsOptional, IsString, ValidateNested } from 'class-validator';
+import { Type } from 'class-transformer';
+import { EntityDto } from 'src/shared/dto/entity.dto';
+import { Fiat } from 'src/shared/models/fiat/fiat.entity';
+
+export class UpdateBankAccountDto {
+  @ApiProperty()
+  @IsNotEmpty()
+  iban: string;
```
Comment (davidleomay) [inline src/payment/models/bank-account/dto/update-bank-account.dto.ts:10]: IBAN cannot be updated

File: src/payment/models/bank-account/dto/update-bank-account.dto.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
+import { IsNotEmpty, IsOptional, IsString, ValidateNested } from 'class-validator';
+import { Type } from 'class-transformer';
+import { EntityDto } from 'src/shared/dto/entity.dto';
+import { Fiat } from 'src/shared/models/fiat/fiat.entity';
+
+export class UpdateBankAccountDto {
```
Comment (davidleomay) [inline src/payment/models/bank-account/dto/update-bank-account.dto.ts:7]: Looks very similar to CreateBankAccountDto => use a common base class?

File: src/payment/models/bank-account/bank-account.controller.ts [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  @ApiBearerAuth()
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.USER))
+  async createBuy(
+    @GetJwt() jwt: JwtPayload,
+    @Body() createBankAccountDto: CreateBankAccountDto,
+  ): Promise<BankAccountDto> {
+    return this.bankAccountService.createBankAccount(jwt.id, createBankAccountDto).then((b) => this.toDto(b));
+  }
+
+  @Put()
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.controller.ts:37]: I would have the the id of the BankAccount to be updated in the route (see e.g. BuyController).

File: src/payment/models/bank-account/bank-account.service.ts [hunk 16]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+
+    const update = this.bankAccountRepo.create(dto);
+
+    Util.removeNullFields(entity);
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:60]: Useless

File: src/payment/models/bank-account/bank-account.service.ts [hunk 17]
```diff
@@ -17,11 +21,50 @@ export class BankAccountService {
...
+  }
+
+  async updateBankAccount(userId: number, dto: UpdateBankAccountDto): Promise<BankAccount> {
+    let entity = await this.bankAccountRepo.findOne({
+      where: { iban: dto.iban, user: { id: userId } },
+      relations: ['user'],
+    });
+    if (!entity) throw new NotFoundException('bankAccount not found');
+
+    const update = this.bankAccountRepo.create(dto);
```
Comment (davidleomay) [inline src/payment/models/bank-account/bank-account.service.ts:58]: We would need to check the currency. Maybe in a separate createEntity method (see e.g. StakingService)?

File: src/payment/models/bank-account/bank-account.controller.ts (hunk 15 above)
Comment (Yannick1712) [reply src/payment/models/bank-account/bank-account.controller.ts:37]: Ok I thought we would work without the id scheme, so for Put still with id?

**Review by davidleomay — APPROVED** [review]

**Pull request merged**